// wechaty-puppet/dist/src/schemas/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MessageType;
(function (MessageType) {
    MessageType[MessageType["Unknown"] = 0] = "Unknown";
    MessageType[MessageType["Attachment"] = 1] = "Attachment";
    MessageType[MessageType["Audio"] = 2] = "Audio";
    MessageType[MessageType["Contact"] = 3] = "Contact";
    MessageType[MessageType["ChatHistory"] = 4] = "ChatHistory";
    MessageType[MessageType["Emoticon"] = 5] = "Emoticon";
    MessageType[MessageType["Image"] = 6] = "Image";
    MessageType[MessageType["Text"] = 7] = "Text";
    MessageType[MessageType["Location"] = 8] = "Location";
    MessageType[MessageType["MiniProgram"] = 9] = "MiniProgram";
    MessageType[MessageType["Money"] = 10] = "Money";
    MessageType[MessageType["Recalled"] = 11] = "Recalled";
    MessageType[MessageType["Url"] = 12] = "Url";
    MessageType[MessageType["Video"] = 13] = "Video";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
/**
 * @private
 */
var WechatAppMessageType;
(function (WechatAppMessageType) {
    WechatAppMessageType[WechatAppMessageType["Text"] = 1] = "Text";
    WechatAppMessageType[WechatAppMessageType["Img"] = 2] = "Img";
    WechatAppMessageType[WechatAppMessageType["Audio"] = 3] = "Audio";
    WechatAppMessageType[WechatAppMessageType["Video"] = 4] = "Video";
    WechatAppMessageType[WechatAppMessageType["Url"] = 5] = "Url";
    WechatAppMessageType[WechatAppMessageType["Attach"] = 6] = "Attach";
    WechatAppMessageType[WechatAppMessageType["Open"] = 7] = "Open";
    WechatAppMessageType[WechatAppMessageType["Emoji"] = 8] = "Emoji";
    WechatAppMessageType[WechatAppMessageType["VoiceRemind"] = 9] = "VoiceRemind";
    WechatAppMessageType[WechatAppMessageType["ScanGood"] = 10] = "ScanGood";
    WechatAppMessageType[WechatAppMessageType["Good"] = 13] = "Good";
    WechatAppMessageType[WechatAppMessageType["Emotion"] = 15] = "Emotion";
    WechatAppMessageType[WechatAppMessageType["CardTicket"] = 16] = "CardTicket";
    WechatAppMessageType[WechatAppMessageType["RealtimeShareLocation"] = 17] = "RealtimeShareLocation";
    WechatAppMessageType[WechatAppMessageType["ChatHistory"] = 19] = "ChatHistory";
    WechatAppMessageType[WechatAppMessageType["MiniProgram"] = 33] = "MiniProgram";
    WechatAppMessageType[WechatAppMessageType["Transfers"] = 2000] = "Transfers";
    WechatAppMessageType[WechatAppMessageType["RedEnvelopes"] = 2001] = "RedEnvelopes";
    WechatAppMessageType[WechatAppMessageType["ReaderType"] = 100001] = "ReaderType";
})(WechatAppMessageType = exports.WechatAppMessageType || (exports.WechatAppMessageType = {}));
/**
 * @private
 */
var WechatMessageType;
(function (WechatMessageType) {
    WechatMessageType[WechatMessageType["Text"] = 1] = "Text";
    WechatMessageType[WechatMessageType["Image"] = 3] = "Image";
    WechatMessageType[WechatMessageType["Voice"] = 34] = "Voice";
    WechatMessageType[WechatMessageType["VerifyMsg"] = 37] = "VerifyMsg";
    WechatMessageType[WechatMessageType["PossibleFriendMsg"] = 40] = "PossibleFriendMsg";
    WechatMessageType[WechatMessageType["ShareCard"] = 42] = "ShareCard";
    WechatMessageType[WechatMessageType["Video"] = 43] = "Video";
    WechatMessageType[WechatMessageType["Emoticon"] = 47] = "Emoticon";
    WechatMessageType[WechatMessageType["Location"] = 48] = "Location";
    WechatMessageType[WechatMessageType["App"] = 49] = "App";
    WechatMessageType[WechatMessageType["VoipMsg"] = 50] = "VoipMsg";
    WechatMessageType[WechatMessageType["StatusNotify"] = 51] = "StatusNotify";
    WechatMessageType[WechatMessageType["VoipNotify"] = 52] = "VoipNotify";
    WechatMessageType[WechatMessageType["VoipInvite"] = 53] = "VoipInvite";
    WechatMessageType[WechatMessageType["MicroVideo"] = 62] = "MicroVideo";
    WechatMessageType[WechatMessageType["SysNotice"] = 9999] = "SysNotice";
    WechatMessageType[WechatMessageType["Sys"] = 10000] = "Sys";
    WechatMessageType[WechatMessageType["Recalled"] = 10002] = "Recalled";
})(WechatMessageType = exports.WechatMessageType || (exports.WechatMessageType = {}));
